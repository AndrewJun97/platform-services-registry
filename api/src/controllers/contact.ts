//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { Request, RequestEditType } from '../db/model/request';
import { fulfillNamespaceEdit } from '../libs/fulfillment';
import shared from '../libs/shared';
import { formatNatsContactObject, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);
const { ContactModel, RequestModel } = dm;
const whichService = 'contact editing';

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    // process request params to get profileId and define RequestEditType
    const { profileId } = params;
    const { productOwner, technicalContact } = body;
    const editType = RequestEditType.Contacts;
    const contacts = [productOwner, technicalContact]

    // Step 1. GET current contact details
    const currentPOvalues = await ContactModel.findById(productOwner.id);
    const currentTCvalues = await ContactModel.findById(technicalContact.id);

    // Step 2. Compare if GithubId or Email values were changed
    const provisionerEdits = [
      currentPOvalues.githubId !== productOwner.githubId,
      currentPOvalues.email !== productOwner.email,
      currentTCvalues.githubId !== technicalContact.githubId,
      currentTCvalues.email !== technicalContact.email,
    ];

    // Step 3. Compare if first or last name details were altered
    const contactNameEdits = [
      currentPOvalues.firstName !== productOwner.firstName,
      currentPOvalues.lastName !== productOwner.lastName,
      currentTCvalues.firstName !== technicalContact.firstName,
      currentTCvalues.lastName !== technicalContact.lastName,
    ];

    // Step 4. Assess if provisioner or contact edits occurred.
    const provisionerEdit = provisionerEdits.some(provisionerEdits => provisionerEdits);
    const contactNameEdit = contactNameEdits.some(contactNameEdits => contactNameEdits);

    if (provisionerEdit) {
      // process request body for natsContext
      const editObject = await formatNatsContactObject(body)
      if (!editObject) {
        const errmsg = 'Cant generate request edit object';
        throw new Error(errmsg);
      }

      const { natsContext, natsSubject } = await fulfillNamespaceEdit(profileId, editType, editObject);

      // create Request record for contact edit
      await RequestModel.create({
        profileId,
        editType,
        editObject: JSON.stringify(contacts),
        natsSubject,
        natsContext: JSON.stringify(natsContext),
      })
      res.status(204).end();
    } else if (contactNameEdit) {
      const updatePromises: any = [];
      contacts.forEach((contact: Contact) => {
        updatePromises.push(ContactModel.update(Number(contact.id), contact))
      })
      await Promise.all(updatePromises);
      res.status(204).end();
    } else {
      res.status(204).end();
    };
  } catch (err) {
  const message = `Unable to update contact`;
  logger.error(`${message}, err = ${err.message}`);
  throw errorWithCode(message, 500);
  }
};

export const processContactEdit = async (request: Request): Promise<void> => {
  try {
    const { editObject } = request;
    const contacts = JSON.parse(editObject);
    const updatePromises: any = [];
    contacts.forEach((contact: Contact) => {
      updatePromises.push(ContactModel.update(Number(contact.id), contact))
    })

    await Promise.all(updatePromises);
    return;
  } catch (err) {
    const message = `Unable to process requestId ${request.id} on bot callback for ${whichService}`;
    logger.error(`${message}, err = ${err.message}`);
    throw err;
  }
};
