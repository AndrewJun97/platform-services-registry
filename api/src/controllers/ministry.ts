//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);
//TODO: Create new ministry api
// export const createMinistry = async (
//   { params, body }: { params: any, body: any }, res: Response
// ): Promise<void> => {
//   const { MinistryModel } = dm;
//   const { profileId } = params;
//   const aBody = { ...body, profileId };

//   const rv = validateObjProps(MinistryModel.requiredFields, aBody);
//   if (rv) {
//     throw rv;
//   }

//   try {
//     const results = await MinistryModel.create(aBody);

//     res.status(201).json(results);
//   } catch (err) {
//     const message = 'Unable create new ministry';
//     logger.error(`${message}, err = ${err.message}`);

//     throw errorWithCode(message, 500);
//   }
// };

export const fetchMinistrySponsors = async (
  { headers, params }: { headers: any, params: any }, res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  // const { profileId } = params;
  // console.log(headers);
  try {
    const results = await MinistryModel.findAllMinistries();


    // const results = [
    //   {
    //     name: 'Citizens\s Services',
    //     code: 'CITZ',
    //   },
    //   {
    //     name: 'Transportation & Infrastructure',
    //     code: 'TRAN',
    //   },
    // ]
    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch ministries`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
