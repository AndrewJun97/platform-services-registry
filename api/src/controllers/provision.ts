//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { fulfillNamespaceProvisioning } from '../libs/fulfillment';
import { MessageType, sendProvisioningMessage } from '../libs/messaging';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);

export const provisionProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { profileId } = params;
  const { ProfileModel, NamespaceModel, ClusterModel } = dm;

  try {
    const existing = await NamespaceModel.findForProfile(profileId);
    if (existing.length === 0) {
      const clusters = await ClusterModel.findAll();
      // TODO:(jl) Everything goes to the default cluster for now.
      const clusterId = clusters.filter(c => c.isDefault === true).pop().id;
      const profile = await ProfileModel.findById(profileId);

      if (!clusterId || !profile) {
        errorWithCode(500, 'Unable to fetch info for provisioning');
      }

      await NamespaceModel.createProjectSet(profileId, clusterId, profile.namespacePrefix);
    }

    await fulfillNamespaceProvisioning(profileId);

    res.status(202).end();
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProvisionedNamespaces = async (
  { body }: { body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel, ClusterModel } = dm;
  const { prefix, clusterName } = body;
  try {

    const namespaces = await NamespaceModel.findForPrefix(prefix);
    const cluster = await ClusterModel.findByName(clusterName);
    logger.info('A');
    if ((!namespaces || namespaces.length === 0) || !cluster) {
      throw new Error();
    }
    logger.info('B');

    const clusterId = cluster.id;
    const provisioned = true;
    const promises: any = [];
    logger.info('C');

    namespaces.forEach(namespace => {
      const { id } = namespace;
      promises.push(NamespaceModel.updateProvisionStatus(Number(id), Number(clusterId), provisioned));
    });
    logger.info('D');

    await Promise.all(promises);
    logger.info('D');

    const profileId = namespaces.pop()!.id;
    // TODO:(jl) This is a catch all endpoint. Needs to be more specific to
    // be used for emailing.
    logger.info('F');

    await sendProvisioningMessage(profileId!, MessageType.ProvisioningCompleted);
    logger.info('G');

    res.status(202).end();
  } catch (err) {
    const message = `Unable to update namespace status`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
