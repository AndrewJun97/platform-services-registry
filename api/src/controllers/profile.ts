//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { QuotaSize } from '../db/model/quota';
<<<<<<< HEAD
import { Request } from '../db/model/request';
import { getQuotaSize } from '../libs/profile';
import { getAllowedQuotaSizes } from '../libs/quota';
import { requestProfileContactsEdit, requestProfileQuotaSizeEdit } from '../libs/request';
import shared from '../libs/shared';
import { validateRequiredFields } from '../libs/utils';
=======
import { getProfileCurrentQuotaSize } from '../libs/profile';
import { getAllowedQuotaSizes } from '../libs/quota';
import { fetchEditRequests, requestContactsEdit, requestQuotaSizeEdit } from '../libs/request';
import shared from '../libs/shared';
>>>>>>> 5fc9710 (refactor and modify unit tests)

const dm = new DataManager(shared.pgPool);

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

<<<<<<< HEAD
    res.status(201).end();
=======
    res.status(204).end();
>>>>>>> 5fc9710 (refactor and modify unit tests)
  } catch (err) {
    const message = `Unable to add contact to profile`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { profileId } = params;

  try {
    const projectContacts = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(projectContacts);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileContacts = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { profileId } = params;
  const { productOwner, technicalContact } = body;
<<<<<<< HEAD
  const contacts = [productOwner, technicalContact];

  // TODO:(yh) add more data sanity check
  // check the passed contacts have no dupliates
  // check contact_id is associated with the queried profile_id
  // check role_id points to the legit role TC / PO
  contacts.forEach((contact: Contact): void => {
    const rv = validateRequiredFields(ContactModel.requiredFields.concat(['id']), contact);
    if (rv) {
      throw rv;
    }
  });

  try {
    const currentContacts: Contact[] = await ContactModel.findForProject(Number(profileId));

    const editCompares: boolean[] = [];
    contacts.forEach((contact: Contact): void => {
      const currentContact = currentContacts.filter(cc => cc.id === contact.id).pop();
      if (!currentContact) {
        throw new Error('Cant get current contact');
      }
      editCompares.push(currentContact.githubId !== contact.githubId);
      editCompares.push(currentContact.email !== contact.email);
    });

    const provisionerRelatedChanges = editCompares.some(editCompare => editCompare);
    if (provisionerRelatedChanges) {
      await requestProfileContactsEdit(Number(profileId), contacts);
      res.status(202).end();
    } else {
      const contactPromises = contacts.map((contact: Contact) => {
        if (!contact.id) {
          throw new Error('Cant get contact id');
        }
        return ContactModel.update(contact.id, contact);
      });
      await Promise.all(contactPromises);
      res.status(204).end();
    }
=======

  try {
    const contacts = [productOwner, technicalContact];
    const currentPOvalues = await ContactModel.findById(productOwner.id);
    const currentTCvalues = await ContactModel.findById(technicalContact.id);

    const editCompares = [
      currentPOvalues.githubId !== productOwner.githubId,
      currentPOvalues.email !== productOwner.email,
      currentTCvalues.githubId !== technicalContact.githubId,
      currentTCvalues.email !== technicalContact.email,
    ];
    const provisionerRelatedChanges = editCompares.some(editCompare => editCompare);

    if (provisionerRelatedChanges) {
      await requestContactsEdit(Number(profileId), body);
    } else {
      const updatePromises: any = [];
      contacts.forEach((contact: Contact) => {
        updatePromises.push(ContactModel.update(Number(contact.id), contact));
      });

      await Promise.all(updatePromises);
    }

    res.status(204).end();
>>>>>>> 5fc9710 (refactor and modify unit tests)
  } catch (err) {
    const message = `Unable to update contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileQuotaSize = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(Number(profileId));
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
>>>>>>> 5fc9710 (refactor and modify unit tests)

    res.status(200).json(quotaSize);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable fetch profile quota size with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileAllowedQuotaSizes = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const allowedQuotaSizes: QuotaSize[] = getAllowedQuotaSizes(quotaSize);

    res.status(200).json(allowedQuotaSizes);
  } catch (err) {
    const message = `Unable to fetch allowed quota-sizes for profile ${profileId}`;
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
    const results = getAllowedQuotaSizes(quotaSize);

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable to fetch quota options for profile ${profileId}`;
>>>>>>> 5fc9710 (refactor and modify unit tests)
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileQuotaSize = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const { requestedQuotaSize } = body;

  try {
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
<<<<<<< HEAD
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const allowedQuotaSizes: QuotaSize[] = getAllowedQuotaSizes(quotaSize);

    // verify if requested quota size is valid
    if (!(requestedQuotaSize && allowedQuotaSizes.includes(requestedQuotaSize))) {
      throw new Error('Please provide correct requested quota size in body');
    }

    await requestProfileQuotaSizeEdit(Number(profileId), requestedQuotaSize);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to update quota-size for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { profileId } = params;

  try {
    const editRequests: Request[] = await RequestModel.findForProfile(Number(profileId));

    res.status(200).json(editRequests);
  } catch (err) {
=======
    const quotaSize: QuotaSize = await getProfileCurrentQuotaSize(profile);
    const allowedQuotaSizes = getAllowedQuotaSizes(quotaSize);

    // verify if requested quota size is valid
    if (!(requestedQuotaSize && allowedQuotaSizes.includes(requestedQuotaSize))) {
      const errmsg = 'Please provide correct requested quota size in body';
      throw new Error(errmsg);
    }

    await requestQuotaSizeEdit(Number(profileId), body);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to request quota size for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { profileId } = params;

  try {
    const results = await fetchEditRequests(profileId);

    res.status(200).json(results);
  } catch (err) {
>>>>>>> 5fc9710 (refactor and modify unit tests)
    if (err.code) {
      throw err;
    }

<<<<<<< HEAD
    const message = `Unable to fetch profile edit requests with profile ID ${profileId}`;
=======
    const message = `Unable fetch requests with profile ID ${profileId}`;
>>>>>>> 5fc9710 (refactor and modify unit tests)
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
