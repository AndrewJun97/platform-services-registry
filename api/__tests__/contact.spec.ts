//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { processContactEdit, requestContactEdit, updateContact } from '../src/controllers/contact';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const selectContact = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/select-contact-edit-request.json');
const selectContactEditRequest = JSON.parse(fs.readFileSync(p1, 'utf8'));

const client = new Pool().connect();

jest.mock('../src/libs/fulfillment', () => {
  const p2 = path.join(__dirname, 'fixtures/provisioning-context.json');
  const natsContext = JSON.parse(fs.readFileSync(p2, 'utf8'));
  const natsSubject = 'edit';
  return {
    fulfillNamespaceEdit: jest.fn().mockResolvedValue({natsContext, natsSubject}),
  };
});

describe('Contact event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();
    // jest.resetAllMocks();

    ex = new FauxExpress();
  });

  it('A contact is updated', async () => {
    const body = {
      firstName: 'John',
      lastName: 'Doe',
      email: 'john@example.com',
      githubId: 'john1100',
    };
    const req = {
      params: { contactId: 233 },
      body,
    }

    client.query.mockResolvedValue({ rows: [selectContact[0]] });

    // @ts-ignore
    await updateContact(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('A contact fails to update', async () => {
    const body = {
      firstName: 'John',
      lastName: 'Doe',
      email: 'john@example.com',
      githubId: 'john1100',
    };
    const req = {
      params: { contactId: 233 },
      body,
    }

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(updateContact(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });

  it('A contact edit request is created', async () => {
    const body = {
      productOwner: {
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'jane1100',
      },
      technicalContact: {
        roleId: 2,
        id: 234,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        githubId: 'john1100',
      },
    };
    const req = {
      params: { profileId: 118 },
      body,
    }

    client.query.mockResolvedValue({
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'jane1100',
    });

    client.query.mockResolvedValue({
      productOwner: {
        userId: 'jane1100',
        provider: 'github',
        email: 'jane.doe@example.ca',
      },
      technicalContact: {
        userId: 'john1100',
        provider: 'github',
        email: 'john.doe@example.com',
      },
    });

    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });

    // @ts-ignore
    await requestContactEdit(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.end).toBeCalled();
  });

  it('A contact edit request fails to be created', async () => {
    const body = {
      productOwner: {
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'jane1100',
      },
      technicalContact: {
        roleId: 2,
        id: 234,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        githubId: 'john1100',
      },
    };
    const req = {
      params: { profileId: 118 },
      body,
    }

    client.query.mockResolvedValue({
      productOwner: {
        userId: 'jane1100',
        provider: 'github',
        email: 'jane.doe@example.com',
      },
      technicalContact: {
        userId: 'john1100',
        provider: 'github',
        email: 'john.doe@example.com',
      },
    });

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(requestContactEdit(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.end).not.toBeCalled();
  });

  it('A contact edit request is processed', async () => {
    const req = selectContactEditRequest

    const editedPO = {
      productOwner: {
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'janedoe',
      },
    };

    const editedTC = {
      technicalContact: {
        roleId: 2,
        id: 234,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        githubId: 'john1100',
      },
    };

    client.query.mockResolvedValue({ ...editedPO, ...editedTC});

    client.query.mockResolvedValueOnce({ rows: [{ editedPO }] });
    client.query.mockResolvedValueOnce({ rows: [{ editedPO }] });
    client.query.mockResolvedValueOnce({ rows: [{ editedTC }] });
    client.query.mockResolvedValueOnce({ rows: [{ editedTC }] });

    JSON.parse = jest.fn().mockImplementationOnce(() => {
      return selectContactEditRequest.editObject;
    });

    await processContactEdit(req);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(JSON.parse).toHaveBeenCalledTimes(1);
  });
});
