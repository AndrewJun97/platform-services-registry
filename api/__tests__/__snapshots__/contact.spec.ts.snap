// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact event handlers A contact edit request is created 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        234,
      ],
    },
  ],
  Array [
    Object {
      "text": "INSERT INTO request
            (profile_id, edit_type, edit_object, nats_subject, nats_context)
            VALUES ($1, $2, $3, $4, $5) RETURNING *;",
      "values": Array [
        118,
        "contacts",
        "[{\\"roleId\\":1,\\"id\\":233,\\"firstName\\":\\"Jane\\",\\"lastName\\":\\"Doe\\",\\"email\\":\\"jane.doe@example.com\\",\\"githubId\\":\\"jane1100\\"},{\\"roleId\\":2,\\"id\\":234,\\"firstName\\":\\"John\\",\\"lastName\\":\\"Doe\\",\\"email\\":\\"john.doe@example.com\\",\\"githubId\\":\\"john1100\\"}]",
        "edit",
        "{\\"profileId\\":118,\\"displayName\\":\\"Project X\\",\\"description\\":\\"This is a sample description\\",\\"namespaces\\":[{\\"namespaceId\\":149,\\"name\\":\\"4ea35c-tools\\",\\"clusters\\":[{\\"clusterId\\":1,\\"name\\":\\"silver\\",\\"provisioned\\":false,\\"quotas\\":{\\"cpu\\":\\"small\\",\\"memory\\":\\"small\\",\\"storage\\":\\"small\\"}}]},{\\"namespaceId\\":151,\\"name\\":\\"4ea35c-test\\",\\"clusters\\":[{\\"clusterId\\":1,\\"name\\":\\"silver\\",\\"provisioned\\":false,\\"quotas\\":{\\"cpu\\":\\"small\\",\\"memory\\":\\"small\\",\\"storage\\":\\"small\\"}}]},{\\"namespaceId\\":150,\\"name\\":\\"4ea35c-dev\\",\\"clusters\\":[{\\"clusterId\\":1,\\"name\\":\\"silver\\",\\"provisioned\\":false,\\"quotas\\":{\\"cpu\\":\\"small\\",\\"memory\\":\\"small\\",\\"storage\\":\\"small\\"}}]},{\\"namespaceId\\":152,\\"name\\":\\"4ea35c-prod\\",\\"clusters\\":[{\\"clusterId\\":1,\\"name\\":\\"silver\\",\\"provisioned\\":false,\\"quotas\\":{\\"cpu\\":\\"small\\",\\"memory\\":\\"small\\",\\"storage\\":\\"small\\"}}]}],\\"productOwner\\":{\\"userId\\":\\"jane1100\\",\\"provider\\":\\"github\\",\\"email\\":\\"jane.doe@example.com\\"},\\"technicalContact\\":{\\"userId\\":\\"john1100\\",\\"provider\\":\\"github\\",\\"email\\":\\"john.doe@example.com\\"}}",
      ],
    },
  ],
]
`;

exports[`Contact event handlers A contact edit request is created 2`] = `204`;

exports[`Contact event handlers A contact edit request is processed 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        234,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE contact
          SET
            first_name = $1, last_name = $2, email = $3, github_id = $4,
            role_id = $5
          WHERE id = 233
          RETURNING *;",
      "values": Array [
        "Jane",
        "Doe",
        "jane@example.com",
        "janedoe",
        1,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE contact
          SET
            first_name = $1, last_name = $2, email = $3, github_id = $4,
            role_id = $5
          WHERE id = 234
          RETURNING *;",
      "values": Array [
        "Jim",
        "Doe",
        "jim@example.com",
        "jim1100",
        2,
      ],
    },
  ],
]
`;

exports[`Contact event handlers A contact is updated 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE contact
          SET
            first_name = $1, last_name = $2, email = $3, github_id = $4,
            role_id = $5
          WHERE id = 233
          RETURNING *;",
      "values": Array [
        "John",
        "Doe",
        "john@example.com",
        "john1100",
        1,
      ],
    },
  ],
]
`;

exports[`Contact event handlers A contact is updated 2`] = `200`;

exports[`Contact event handlers A contact is updated 3`] = `
Object {
  "archived": false,
  "createdAt": "2020-09-10T18:14:13.436Z",
  "email": "jane@example.com",
  "firstName": "Jane",
  "githubId": "jane1100",
  "id": 233,
  "lastName": "Doe",
  "roleId": 1,
  "updatedAt": "2020-09-10T18:14:13.436Z",
}
`;
